Modernize CMake scripts

roehling opened the pull request:
This is a rewrite of the CMake configuration that removes a lot of legacy code which is no longer needed and makes the build system more convoluted than necessary.

By switching to modern CMake semantics, we gain the following advantages:

* Include directories and dependent libraries are associated with the exported target, no additional configuration beyond target_link_libraries() required.

* Built-in Qt MOC and UIC handling

* Proper separation of build and install interface; this also makes RPATH just work out of the box without additional configuration.

* Native build of vendored QGLViewer library, no need to call qmake and import the result.

* Proper handling and versioning of dependencies. No hard-coded build paths showing up in unexpected places.

ahornung commented:
Thanks a lot, that seems long overdue! I haven't had a chance to look at all the changes yet, but could this also solve some of the other issues related to CMake (#392  #321 #225 #223 #270 )?

Ping @wxmerkt @mwoehlke-kitware @SpaceIm  just to get some additional input.

roehling commented:
I believe this might actually solve all the issues you listed. :)

Btw, I just force-pushed the branch because there was a typo in the commit message and I also forgot a closing parenthesis in the octovis config script (which only causes an issue  if the system QGLViewer is used). I'll gladly fix any other bugs you encounter. I also made some design choices which I think are good but YMMV. I'm open to suggestions if you think something should be solved differently.

ahornung commented:
@roehling, sounds great!

I just merged some smaller cleanups and fixed CI calls into 'devel', could you rebase on these changes?

roehling committed:
Modernize CMake scripts

This is a rewrite of the CMake configuration that removes a lot of
legacy code which is no longer needed and makes the build system more
convoluted than necessary.

By switching to modern CMake semantics, we gain the following
advantages:

* Include directories and dependent libraries are associated with the
exported target, no additional configuration beyond
target_link_libraries() required.

* Built-in Qt MOC and UIC handling

* Proper separation of build and install interface; this also makes
RPATH just work out of the box without additional configuration.

* Native build of vendored QGLViewer library, no need to call qmake and
import the result.

* Proper handling and versioning of dependencies. No hard-coded build
paths showing up in unexpected places.

Signed-off-by: Timo Röhling <timo.roehling@fkie.fraunhofer.de>

roehling committed:
Update vendored copy of QGLViewer to version 2.9.1

This is needed for Qt6 support.

Signed-off-by: Timo Röhling <timo.roehling@fkie.fraunhofer.de>

roehling commented:
Done. I also updated QGLViewer to the latest release 2.9.1, because the currently vendored copy does not support Qt6.

ahornung commented on CMakeLists.txt:
```diff
@@ -1,21 +1,22 @@
+cmake_minimum_required(VERSION 3.10...4.1)
+project(
+    octomap-distribution
+    VERSION 1.10.0
```
If I get it right, the package version is set here and then used later on (instead of the previous MAJOR/MINOR/PATCH variables)? We would then also need an update scripts/increase_version.py - or change back to single variables, which could be easier to change.

roehling replied:
I tried to keep the different subprojects self-contained, which is why I avoided using settings from the super project. I will update the helper script if that is okay for you.

ahornung replied:
Sure thing, more than OK!

ahornung commented on dynamicEDT3D/CMakeLists.txt:
```diff
@@ -1,177 +1,139 @@
+cmake_minimum_required(VERSION 3.10...4.1)
+project(
+    dynamicEDT3D
+    VERSION 1.10.0
```
See first comment on VERSION

ahornung commented on octomap/CMakeLists.txt:
```diff
@@ -1,180 +1,216 @@
+cmake_minimum_required(VERSION 3.10...4.1)
+project(
+    octomap
+    VERSION 1.10.0
```
See first comment on VERSION

ahornung commented on octovis/CMakeLists.txt:
```diff
@@ -1,168 +1,265 @@
+cmake_minimum_required(VERSION 3.10...4.1)
+project(
+    octovis
+    VERSION 1.10.0
```
See first comment on VERSION

ahornung commented on dynamicEDT3D/CMakeLists.txt:
```diff
@@ -1,177 +1,139 @@
-# Install catkin package.xml, attention package.xml names the catkin package "dynamic_edt_3d", so this is also the location where it needs to be installed to (and not "dynamicEDT3D")
-install(FILES package.xml DESTINATION "${CMAKE_INSTALL_DATADIR}/dynamic_edt_3d")
+add_library(dynamicEDT3D ${dynamicEDT3D_SOURCES})
+target_compile_features(dynamicEDT3D PUBLIC cxx_std_11)
```
I think we can safely get rid of that line, no need to pin to C++11 (anymore).

roehling replied:
Unlike the CXX_STANDARD property, the compile feature does not pin the version, it merely sets the minimum requirement. I would recommend to leave it in as a way to document the minimum required C++ standard.

ahornung commented on octomap/CMakeLists.txt:
```diff
@@ -1,180 +1,216 @@
+                     src/math/Vector3.cpp
 )
+add_library(octomath ${octomath_SOURCES})
+target_compile_features(octomath PUBLIC cxx_std_11)
```
I think we can safely get rid of that line, no need to pin to C++11 (anymore).

roehling replied:
[see above](https://github.com/OctoMap/octomap/pull/439#discussion_r2449604000)

ahornung commented on octomap/CMakeLists.txt:
```diff
@@ -1,180 +1,216 @@
+    src/ColorOcTree.cpp
+)
+add_library(octomap ${octomap_SOURCES})
+target_compile_features(octomap PUBLIC cxx_std_11)
```
I think we can safely get rid of that line, no need to pin to C++11 (anymore).

roehling replied:
[see above](https://github.com/OctoMap/octomap/pull/439#discussion_r2449604000)

ahornung commented on octovis/CMakeLists.txt:
```diff
@@ -1,168 +1,265 @@
+    include/octovis/ViewerWidget.h
+)
+add_library(octovis ${octovis_HEADERS} ${octovis_SOURCES})
+target_compile_features(octovis PUBLIC cxx_std_11)
```
I think we can safely get rid of that line, no need to pin to C++11 (anymore).

roehling replied:
[see above](https://github.com/OctoMap/octomap/pull/439#discussion_r2449604000)

ahornung commented:
> Done. I also updated QGLViewer to the latest release 2.9.1, because the currently vendored copy does not support Qt6.

Perfect, now all CI checks are green as well (and the windows one will probably as well once re-activated - since #321 might be fixed).

Regarding the latest change added for Qt, I'm thinking about setting Qt5 as default (#442), depending on what's mostly used by the consumers e.g. in ROS distros and what creates fewer issues or hiccups. Using the system-provided QGLViewer should be preferred. Thoughts on this?

ahornung commented:
Also friendly ping to @clalancette, just in case there is input from your side to consider regarding ROS integration.

roehling commented:
Taking the external QGLViewer library Qt version into consideration is a good point. I have an idea how to do that, I'll push an update later.

roehling committed:
Try harder to pick matching Qt and QGLViewer libraries

Signed-off-by: Timo Röhling <timo.roehling@fkie.fraunhofer.de>

roehling commented:
So I've added some code to detect the QQLViewer Qt version (very naively from the library filename) and turned the Qt version selector option into a tri-state ("auto", "5", "6"). By default, octovis now looks for the system QGLViewer library. If found, it forces a matching Qt version. If not found, it will build QGLViewer from source and pick Qt 5 over Qt 6.

The pure CMake CI build jobs started failing because I am now treating all missing dependencies as errors. Before, the octovis build was silently skipped.

ahornung commented:
@roehling, thanks, sounds useful! I think when building the whole distribution, the previous behavior of skipping octovis if the dependencies were missing would be preferable, since it's just an additional tool with rather large set of additional dependencies. I'll also try to add CI build for the single packages, just in case.

Note that I also left a few code review comments above, not sure if you got a notification about them.

roehling commented:
> @roehling, thanks, sounds useful! I think when building the whole distribution, the previous behavior of skipping octovis if the dependencies were missing would be preferable, since it's just an additional tool with rather large set of additional dependencies. I'll also try to add CI build for the single packages, just in case.

I can revert that change.

> Note that I also left a few code review comments above, not sure if you got a notification about them.

No, I did not, and I can't seem to find them at all, sorry.

ahornung reviewed (changes requested):
@roehling testing the review feature, please check comments in the review.

ahornung commented:
> > Note that I also left a few code review comments above, not sure if you got a notification about them.
> 
> No, I did not, and I can't seem to find them at all, sorry.

My bad, apparently the comments are drafts and only visible to me, until I click "Submit review". Are they now visible here or in the changed files?

roehling committed:
Add option to make octovis build optional

Signed-off-by: Timo Röhling <timo.roehling@fkie.fraunhofer.de>

roehling committed:
Fix increase_version.py to deal with new CMakeLists.txt files

Signed-off-by: Timo Röhling <timo.roehling@fkie.fraunhofer.de>

roehling committed:
Add missing return()

Signed-off-by: Timo Röhling <timo.roehling@fkie.fraunhofer.de>

ahornung reviewed (approved):
@roehling  Thanks for your contributions, you even greatly improved the version bump helper!

Ready to merge? From my side all fine!
